File: ruby/greeter/lib/greeter.rb
class InputHandler
  def self.get_user_name
    print "Enter your name: "
    user_input = gets.chomp

    name = user_input.strip

    name = "Stranger" if name.empty?

    name
  end
end

class Greeter
  def initialize(name)
    @username = name
  end

  def greet
    puts "Hello, #{@username}!"
    puts "Greetings from Ruby!"
  end
end

def main
  name = InputHandler.get_user_name
  greeter = Greeter.new(name)

  greeter.greet
end

main

File: elixir/greeter/lib/greeter.ex
defmodule InputHandler do
  def get_user_name do
    IO.gets("Enter your name: ")
    |> String.trim()
    |> case do
      "" -> "Stranger"
      name -> name
    end
  end
end

defmodule Greeter do
  defstruct username: ""

  def create(name) do
    %Greeter{username: name}
  end

  def greet(%Greeter{username: name}) do
    IO.puts("Hello, #{name}!")
    IO.puts("Greetings from Elixir!")
  end

  def start(_) do
    name = InputHandler.get_user_name()
    greeter = Greeter.create(name)

    Greeter.greet(greeter)
  end
end
